server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java

```diff
@@ -142,6 +145,34 @@ public void testCancelException() {
         assertPropagatesInterrupt(() -> future.actionGet(10, TimeUnit.SECONDS));
     }
 
+    public void testAssertCompleteAllowedAllowsConcurrentCompletesFromSamePool() {
+        assumeTrue("Assertions need to be enabled for assertCompleteAllowed to be invoked", Assertions.ENABLED);
+        final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());
+        final var executorName = randomFrom(ThreadPool.Names.GENERIC, ThreadPool.Names.MANAGEMENT);
+        final var threadCount = 4;
+        final var running = new AtomicBoolean(true);
+        final var startBarrier = new CyclicBarrier(threadCount + 1);
+        try (TestThreadPool threadPool = new TestThreadPool(getTestName())) {
+            assert threadPool.info(executorName).getMax() >= threadCount : "Not enough threads in pool";
+            // N threads competing to complete the futures
+            for (int i = 0; i < threadCount; i++) {
+                threadPool.executor(executorName).submit(() -> {
+                    safeAwait(startBarrier);
+                    while (running.get()) {
+                        futureReference.get().onResponse(null);
+                    }
+                }, null);
+            }
+            // create new futures to complete
```

Comment:
Could you expand this comment to describe why we're creating multiple futures, recording the investigations you've done in this area?